src/main/assisted-installer-controller/assisted_installer_main.go

```diff
@@ -55,38 +62,77 @@ func main() {
 		kc,
 	)
 
+	assistedController.SetReadyState()
+
 	// While adding new routine don't miss to add wg.add(1)
 	// without adding it will panic
 	var wg sync.WaitGroup
-	var wgLogs sync.WaitGroup
 	var status assistedinstallercontroller.ControllerStatus
 
-	ctxApprove, cancelApprove := context.WithCancel(context.Background())
-	go assistedController.ApproveCsrs(ctxApprove, &wg)
+	ctxRoutines, cancelRoutines := context.WithCancel(context.Background())
+	go assistedController.WaitAndUpdateNodesStatus(ctxRoutines, &wg)
 	wg.Add(1)
-	go assistedController.PostInstallConfigs(&wg, &status)
+	go assistedController.PostInstallConfigs(ctxRoutines, &wg, &status)
 	wg.Add(1)
-	go assistedController.UpdateBMHs(&wg)
+	go assistedController.UpdateBMHs(ctxRoutines, &wg)
 	wg.Add(1)
+
+	// No need to cancel with context, will finish quickly
 	go assistedController.HackDNSAddressConflict(&wg)
 	wg.Add(1)
 
-	ctxLogs, cancelLogs := context.WithCancel(context.Background())
-	go assistedController.UploadLogs(ctxLogs, cancelLogs, &wgLogs, &status)
-	wgLogs.Add(1)
+	go assistedController.UploadLogs(ctxRoutines, &wg, &status)
+	wg.Add(1)
 
-	assistedController.SetReadyState()
-	assistedController.WaitAndUpdateNodesStatus(&status)
-	logger.Infof("Sleeping for 10 minutes to give a chance to approve all csrs")
-	time.Sleep(10 * time.Minute)
-	cancelApprove()
+	waitForInstallation(client, logger, &status)
+	// stop all go routines
+	cancelRoutines()
 
 	logger.Infof("Waiting for all go routines to finish")
 	wg.Wait()
-	if !status.HasError() {
-		//with error the logs are canceled within UploadLogs
-		logger.Infof("closing logs...")
-		cancelLogs()
+	logger.Infof("Finished all")
+}
+
+func waitForInstallation(client inventory_client.InventoryClient, log logrus.FieldLogger, status *assistedinstallercontroller.ControllerStatus) {
+	log.Infof("monitor cluster installation status")
+	reqCtx := utils.GenerateRequestContext()
+	errCounter := 0
+	for {
+		time.Sleep(assistedinstallercontroller.GeneralWaitInterval)
+		cluster, err := client.GetCluster(reqCtx)
+		if err != nil {
+			// In case cluster was deleted or controller is not authorised
+			// we should exit controller
+			switch err.(type) {
+			case *installer.GetClusterNotFound:
+				errCounter++
+				log.WithError(err).Errorf("Cluster was not found in inventory or user is not authorized")
+			case *installer.GetClusterUnauthorized:
```

Comment:
I recommend moving this to a retry utility or use a 3-rd party library

Reply:
Not sure it is needed in this case. It is more visible what is going on. Another way i will need to pack in function and use it with retry package. I didn't like it, so i used this approach

Reply:
As the reader of the code it was actually less visible. At least add detailed comment about the need and the reason for trying 10 times

Reply:
I think "for" already implement retry logic, not sure that it is better to have retry inside retry .